internal/reconcile/deployment.go

```diff
@@ -0,0 +1,58 @@
+package reconcile
+
+import (
+	"context"
+	"fmt"
+
+	log "github.com/ViaQ/logerr/v2/log/static"
+	"github.com/openshift/cluster-logging-operator/internal/constants"
+	util "github.com/openshift/cluster-logging-operator/internal/utils"
+	"github.com/openshift/cluster-logging-operator/internal/utils/comparators/pod"
+	apps "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Deployment reconciles a Deployment to the desired spec returning an error
+// if there is an issue creating or updating to the desired state
+func Deployment(er record.EventRecorder, k8Client client.Client, desired *apps.Deployment) error {
+	reason := constants.EventReasonGetObject
+	updateReason := ""
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		current := &apps.Deployment{}
+		key := client.ObjectKeyFromObject(desired)
+		if err := k8Client.Get(context.TODO(), key, current); err != nil {
+			if errors.IsNotFound(err) {
+				reason = constants.EventReasonCreateObject
+				return k8Client.Create(context.TODO(), desired)
+			}
+			return fmt.Errorf("failed to get %v Deployment: %w", key, err)
+		}
+		same := false
+
+		if same, updateReason = pod.AreSame(&current.Spec.Template.Spec, &desired.Spec.Template.Spec, current.Name); same && util.HasSameOwner(current.OwnerReferences, desired.OwnerReferences) {
```

Comment:
This is missing something as it only compares the template spec but Deployments and Daemonsets have other differences we may wish to control, specifically replicas and labels?  @alanconway can you weigh in here.  The way this is implemented admins control this field which is a departure from other objects, but "replicacount" would mean we would need to provide some other mechanism of control

Reply:
There is no regression in what we had before. We were spinning through the pods and comparing them. There were no other specs specific to daemonsets that were compared.

See: https://github.com/openshift/cluster-logging-operator/pull/2240#discussion_r1385685305 also https://github.com/openshift/cluster-logging-operator/pull/2240#discussion_r1388502398

Reply:
Since we are updating the labels below, I think we should be comparing them above - we should compare the same set of values that you want to update.

Reply:
I've added back the daemonset comparator and added a deployment comparator. These compare the pod specs along with labels and owners.